### PR TITLE
mgmt, support encryption with user-assigned managed identity

### PR DESCRIPTION
**Pull request by XiaofeiCao**
# Description

fix https://github.com/Azure/azure-sdk-for-java/issues/38854

doc: https://learn.microsoft.com/en-us/azure/storage/common/customer-managed-keys-overview

Also added an overload for updating all attached resources of a VM, code: https://github.com/Azure/azure-sdk-for-java/pull/38937/commits/fa185a4c96f0c4d3c1f945665656debeecd0bfaf

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java [hunk 1]
```diff
@@ -442,6 +501,14 @@ interface WithUserAssignedManagedServiceIdentity {
              * @return the next stage of the storage account definition
              */
             WithCreate withExistingUserAssignedManagedServiceIdentity(Identity identity);
+
+            /**
+             * Specifies an existing user assigned identity to be associated with the storage account.
+             *
+             * @param identityId ID of the identity
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:508]: nit, "the resource ID of the user-assigned maanged service identity"?

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java [hunk 2]
```diff
@@ -746,6 +814,51 @@ interface WithEncryption {
...
+             * @return the next stage of storage account update
+             */
+            Update withEncryptionKeyFromKeyVault(String keyVaultUri, String keyName, String keyVersion, String userAssignedIdentityId);
+
+            /**
+             * Specifies the Microsoft-managed key to be used as key for encryption. This is the default encryption type.
+             *
+             * @return the next stage of storage account update
+             */
+            Update withEncryptionKeyFromStorage();
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:861]: A bit unsure about the name. candidate: "withEncryptionKeyFromMicrosoft" "withMicrosoftManagedEncryptionKey"

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:861]: I'm not sure either. We had a method `encryptionKeySource()` exposed, with two possible values `Microsoft.Storage` and `Microsoft.Keyvault`.

Current relations are:
* `withEncryptionKeyFromKeyVault` -> `Microsoft.Keyvault`
* `withEncryptionKeyFromStorage`(`withMicrosoftManagedEncryptionKey`) -> `Microsoft.Storage`

Guess `withMicrosoftManagedEncryptionKey` should be fine?

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:861]: Yeah, personally I am on `withMicrosoftManagedEncryptionKey`

@haolingdong-msft  your opinion as well?

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java [hunk 3]
```diff
@@ -226,6 +226,27 @@ public interface StorageAccount
...
 
+    /**
+     * Type of the Managed Service Identity used to access KeyVault for encryption.
+     * <p>This property only makes sense when the storage account is encrypted using Customer-managed keys, meaning
+     * {@link StorageAccount#encryptionKeySource()} is {@link StorageAccountEncryptionKeySource#MICROSOFT_KEYVAULT}.</p>
+     *
+     * @return type of the Managed Service Identity used to access KeyVault for encryption,
+     *         null if {@link StorageAccount#encryptionKeySource()} is {@link StorageAccountEncryptionKeySource#MICROSOFT_STORAGE}
+     */
+    IdentityType identityTypeForKeyVault();
```
[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:237]: Maybe `identityTypeForCustomerManagedEncryptionKey`

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java (hunk 2 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:861]: @haolingdong-msft 

`withEncryptionKeyFromKeyVault` -> encryption using Customer Managed Keys
`withEncryptionKeyFromStorage(withMicrosoftManagedEncryptionKey)` -> encryption using Microsoft Managed Keys

doc: [withEncryptionKeyFromStorage(withMicrosoftManagedEncryptionKey)](https://learn.microsoft.com/en-us/azure/storage/common/customer-managed-keys-overview)

Comment (haolingdong-msft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:861]: My gust feeling is the signature would be better to be: `withEncryptionKey(EncryptionKeyKindEnum)`. But as we already exposed `withEncryptionKeyFromKeyVault()`, I think it's impossible to change to`withEncryptionKey(EncryptionKeyKindEnum)`.

I don't have strong prereference for this. Maybe `withMicrosoftManagedEncryptionKey` to better align with the document.

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java]: `EncryptionKeyKindEnum` probably won't work, as MMK and CMK be very different on required configs.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java [hunk 4]
```diff
@@ -226,6 +232,27 @@ public interface StorageAccount
      */
     boolean isDefaultToOAuthAuthentication();
 
+    /**
+     * Type of the Managed Service Identity used to access KeyVault for encryption.
+     * <p>This property only makes sense when the storage account is encrypted using Customer-managed keys, meaning
+     * {@link StorageAccount#encryptionKeySource()} is {@link StorageAccountEncryptionKeySource#MICROSOFT_KEYVAULT}.</p>
+     *
+     * @return type of the Managed Service Identity used to access KeyVault for encryption,
+     *         null if {@link StorageAccount#encryptionKeySource()} is {@link StorageAccountEncryptionKeySource#MICROSOFT_STORAGE}
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:241]: nit, `{@code null}`?

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java [hunk 5]
```diff
@@ -381,6 +408,44 @@ interface WithEncryption {
              * @return the next stage of storage account definition
              */
             WithCreate withEncryptionKeyFromKeyVault(String keyVaultUri, String keyName, String keyVersion);
+
+            /**
+             * Specifies the KeyVault key to be used as key for encryption and the user-assigned identity to access the KeyVault,
+             * meaning this Storage Account will be encrypted with Customer-managed keys.
+             * If the provided user-assigned identity is null, this method is equivalent to {@link WithEncryption#withEncryptionKeyFromKeyVault(String, String, String)},
+             * meaning it'll use the system-assigned identity to access the KeyVault.
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:416]: Personally a bit unsure about whether we should "require" user sets a non-null identity.

On one hand it better to require non-null given the API shape.

On the other hand I can think of some case that putting a nullable identity would be simpler than conditioning (and fluent style is hard to do a `if else`).

So maybe keep this.

**Review by weidongxu-microsoft — APPROVED** [review]
As this is a pretty big one, please get approval from @haolingdong-msft as well.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java (hunk 5 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:416]: I get your point. The meaning of putting a null identity here is a bit vague to the user, even with the javadocs. User could accidentally provide a null identity and the behavior may surprise them. We should require a non-null identity.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageEncryptionHelper.java [hunk 6]
```diff
@@ -175,14 +192,25 @@ StorageEncryptionHelper withInfrastructureEncryption() {
...
         encryption.withKeySource(KeySource.MICROSOFT_KEYVAULT);
         encryption
+            .withEncryptionIdentity(new EncryptionIdentity().withEncryptionUserAssignedIdentity(identityId))
             .withKeyVaultProperties(
                 new KeyVaultProperties().withKeyVaultUri(keyVaultUri).withKeyName(keyName).withKeyVersion(keyVersion));
         return this;
     }
 
+    public StorageEncryptionHelper withEncryptionKeyFromStorage() {
+        Encryption encryption = getEncryptionConfig(true);
```
Comment (haolingdong-msft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageEncryptionHelper.java:209]: do we want to rename this method as well?

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageEncryptionHelper.java:209]: This is in implementation and not exposed to user. Guess it's fine. Personally I prefer staying with `withEncryptionKeyFromStorage` to describe how it's implemented.

**Review by haolingdong-msft — APPROVED** [review]

**Pull request merged**